### PR TITLE
feat: operator support for env. variables

### PR DESCRIPTION
**Pull request by jsenko**

### CODE REVIEW

**Comment by jsenko** [comment]
hi @andreaTP take a look please when you have time

File: operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        return this;
+    }
+
+    private void recursiveDependencyOrdering(List<EnvVar> res, EnvCacheEntry current, int depth) {
+        if (depth > env.size()) {
+            // spotless:off
+            throw new OperatorException("""
+                    Cycle detected during the processing of environment variables (at %s), make sure that every env. variable is define once. \
+                    Explanation: Ordering of the env. variables is significant, because some variables can reference others using interpolation. \
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java:61]: Ok, I think this is the problem we are trying to solve here.
Instead of trying hard to solve the problem for the user, let's just use, the env vars passed by the user, honoring the order.
This way, a user always has a mechanism to get interpolations working without us trying to fix it.

We shouldn't try to track cycles, that will be user misconfiguration.

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks for starting this effort, I believe that we can use a dumbest approach.

Happy to refine in a meeting the problem further if I'm missing any major use case that's not covered by the comments.

File: operator/controller/src/main/java/io/apicurio/registry/operator/ApicurioRegistry3Reconciler.java [hunk 2]
```diff
@@ -63,16 +63,26 @@ public class ApicurioRegistry3Reconciler implements Reconciler<ApicurioRegistry3
 
     public UpdateControl<ApicurioRegistry3> reconcile(ApicurioRegistry3 primary,
             Context<ApicurioRegistry3> context) {
+        try (var ru = new ResourceUtils<>(primary, context, REGISTRY_KEY)) {
 
-        log.info("Reconciling Apicurio Registry: {}", primary);
-        var statusUpdater = new StatusUpdater(primary);
+            var statusUpdated = cell(false);
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/ApicurioRegistry3Reconciler.java:68]: Values used in lambda expressions have to be effectively final. This allows to mutate a value in and outside of lambdas. I think the `ru.withDesiredResource` function is better that repeatedly checking for optional result from `context.getSecondaryResource`, and they nest nicely.

File: operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import io.fabric8.kubernetes.api.model.EnvVarBuilder;
+
+import java.util.*;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class manages a collection of env. variables that should be applied to a Deployment.
+ */
+public class EnvCache {
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java:14]: It solves the following problems:
- Keeps track of env. variables in a map. Do you prefer to have it as a field in each resource? This way the logic is encapsulated and can be passed to other methods and object. Most of what the operator will do is to manipulate env. variables.
- It handles the precedence, so env. variables set by the operator are not overwriting user-defined ones.
- It keeps the ordering of the variables as user defined them in the spec.

File: operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java (hunk 1 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java:61]: This logic makes sure that the ordering is kept for variables that come from the spec.tries to order the variable based on the dependency field. Do you want me to use LinkedHashMap?

File: operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCacheEntry.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Getter
+@EqualsAndHashCode
+@ToString
+public class EnvCacheEntry {
+
+    private final EnvVar var;
+
+    private final EnvCachePriority priority;
+
+    private final List<String> dependencies;
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCacheEntry.java:21]: No it does not track dependencies based on content. This field contains names of variables that should be ordered before this one. As simple as that.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java [hunk 5]
```diff
@@ -33,25 +33,31 @@ public AppDeploymentResource() {
 
     @Override
     protected Deployment desired(ApicurioRegistry3 primary, Context<ApicurioRegistry3> context) {
+        try (var ru = new ResourceUtils<>(primary, context, APP_DEPLOYMENT_KEY)) {
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java:36]: Have you seen what the `close` method does? I'll remove the MDC completely and let someone else handle it later.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Cell.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package io.apicurio.registry.operator.utils;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+/**
+ * Non-atomic variable wrapper that can be final and can be used to share data with lambda functions.
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/Cell.java:9]: Responded in the first comment.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/TraverseUtils.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public class TraverseUtils {
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/TraverseUtils.java:11]: It's a utility method to do things like this:
```
where(d.getSpec().getTemplate().getSpec().getContainers(), c -> APP_CONTAINER_NAME.equals(c.getName()), 
   c -> {
        c.setEnv(ec.getEnv());
});
```
It has nothing to do with dependent resources.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+                ApicurioRegistry3.class);
+        registry.getMetadata().setNamespace(namespace);
+        registry.getSpec().getApp().setHost(ingressManager.getIngressHost("app"));
+        registry.getSpec().getUi().setHost(ingressManager.getIngressHost("ui"));
+
+        client.resource(registry).create();
+
+        await().ignoreExceptions().until(() -> {
+
+            var appEnv = whereR(client.apps().deployments().inNamespace(namespace).withName(registry.getMetadata().getName() + "-app-deployment").get().getSpec().getTemplate().getSpec().getContainers(), c -> APP_CONTAINER_NAME.equals(c.getName()), Container::getEnv);
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java:46]: So, do you want me to replace these with explicit loops instead of a helper functions? Sure I can write a bunch of loops.

**Comment by jsenko** [comment]
I'll update this code to be as simple as possible if that's what you require. None (maybe one) of the comments were about correctness or logical issues with the code, just code style. Moreover, three of the comments looked like you didn't take time to understand what the code is trying to do, just skimmed the PR. But I'll trim it down so we can move on, and let someone else worry about refactoring later. I think we agree that we both want to get this over with.

File: operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/env/EnvCache.java:14]: An `ArrayList` should be able to satisfy the requirement.

**Comment by jsenko** [comment]
@andreaTP updated

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java [hunk 9]
```diff
@@ -36,22 +38,36 @@ protected Deployment desired(ApicurioRegistry3 primary, Context<ApicurioRegistry
...
+        addEnvVar(envVars, new EnvVarBuilder().withName("QUARKUS_HTTP_ACCESS_LOG_ENABLED").withValue("true").build());
+        addEnvVar(envVars, new EnvVarBuilder().withName("QUARKUS_HTTP_CORS_ORIGINS").withValue("*").build());
+        addEnvVar(envVars, new EnvVarBuilder().withName("APICURIO_REST_DELETION_GROUP_ENABLED").withValue("true").build());
+        addEnvVar(envVars, new EnvVarBuilder().withName("APICURIO_REST_DELETION_ARTIFACT_ENABLED").withValue("true").build());
+        addEnvVar(envVars, new EnvVarBuilder().withName("APICURIO_REST_DELETION_ARTIFACTVERSION_ENABLED").withValue("true").build());
+        addEnvVar(envVars, new EnvVarBuilder().withName("APICURIO_APIS_V2_DATE_FORMAT").withValue("yyyy-MM-dd''T''HH:mm:ssZ").build());
+        // spotless:on
+
+        // This must be done after any modification of the map by the operator.
+        primary.getSpec().getApp().getEnv().forEach(e -> {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java:55]: I find it a bit hard to follow the logic, what do you think about building directly the final map?

Something like:
```java
var envVars = new LinkedHashMap<String, EnvVar>();
for (var e: primary.getSpec().getApp().getEnv()) {
    envVars.put(e.getName(), e);
}
```
and change `addEnvVar` by adding:
```java
if (!map.containsKey(envVar.getName())) {
```

am I missing any invariant?

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            for (var c : appContainers) {
+                if (APP_CONTAINER_NAME.equals(c.getName())) {
+                    appEnv = c.getEnv();
+                }
+            }
+            assertThat(appEnv).map(EnvVar::getName).containsOnlyOnce(defaultAppEnv);
+            var QUARKUS_HTTP_ACCESS_LOG_ENABLED = appEnv.stream().filter(e -> "QUARKUS_HTTP_ACCESS_LOG_ENABLED".equals(e.getName())).map(EnvVar::getValue).findAny().orElse(null);
+            assertThat(QUARKUS_HTTP_ACCESS_LOG_ENABLED).isEqualTo("false");
+            assertThat(appEnv).containsSubsequence(
+                    new EnvVarBuilder().withName("APP_VAR_1_NAME").withValue("APP_VAR_1_VALUE").build(),
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java:103]: nitpick: maybe worth building those `EnvVar`s outside the test body to improve readability?

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            "APICURIO_REST_DELETION_ARTIFACT_ENABLED",
+            "APICURIO_REST_DELETION_ARTIFACTVERSION_ENABLED",
+            "APICURIO_APIS_V2_DATE_FORMAT"
+    };
+    // spotless:on
+
+    private static final String[] defaultUIEnv = new String[] { "REGISTRY_API_URL" };
+
+    @Test
+    void testEnvVars() {
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java:40]: nitpick: I think that it will read a bit better if we split this single test in multiple, shorter once, wdyt?

**Review by andreaTP — COMMENTED** [review]
Feel free to ignore the nitpicks if that's too much for you, added a couple of suggestions.
Thanks for turning this around!

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java (hunk 11 above)
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/EnvITTest.java:41]: Addressing this will be very welcome, but not a blocker.

**Review by andreaTP — APPROVED** [review]
lgtm, thanks!
